mirascope/mcp/client.py

```diff
@@ -0,0 +1,280 @@
+            )
+        else:
+            fields[prop_name] = (
+                type_ | None,
+                Field(None, description=prop_schema.get("description")),
+            )
+
+    return create_model(name, __config__=ConfigDict(extra="allow"), **fields)
+
+
+def json_schema_to_python_type(schema: dict[str, Any], name: str = "Model") -> Any:
+    """
+    Recursively convert a JSON Schema snippet into a Python type annotation or a dynamically generated pydantic model.
+    """
+    json_type = schema.get("type", "any")
+
+    if json_type == "string":
+        return str
+    elif json_type == "number":
+        return float
+    elif json_type == "integer":
+        return int
+    elif json_type == "boolean":
+        return bool
+    elif json_type == "array":
+        # Recursively determine the items type for arrays
+        items_schema = schema.get("items", {})
+        items_type = json_schema_to_python_type(items_schema)
+        return list[items_type]
+    elif json_type == "object":
+        # Recursively build a dynamic model for objects
+        properties = schema.get("properties", {})
+        required = schema.get("required", [])
+        return build_object_type(properties, required, name)
+    else:
+        # Default fallback
+        return Any
+
+
+def create_model_from_tool(tool: Tool) -> type[BaseModel]:
+    """
+    Create a dynamic pydantic model from the given Tool instance based on its inputSchema.
+    """
+    schema = tool.inputSchema
+    properties = schema.get("properties", {})
+    required_fields = schema.get("required", [])
+
+    fields = {}
+    for field_name, field_schema in properties.items():
+        field_type = json_schema_to_python_type(field_schema)
+
+        if field_name in required_fields:
+            default = Field(..., description=field_schema.get("description"))
+            annotation = field_type
+        else:
+            default = Field(None, description=field_schema.get("description"))
+            annotation = field_type | None
+
+        fields[field_name] = (annotation, default)
+
+    return create_model(
+        snake_to_pascal(tool.name), __config__=ConfigDict(extra="allow"), **fields
+    )
+
+
+def _create_model_from_prompt(prompt: Prompt) -> type[BaseModel]:
+    if not prompt.arguments:
+        return create_model(prompt.name, __base__=MCPClientTool)
+
+    fields = {}
+    for arg in prompt.arguments:
+        if arg.required:
+            default = Field(..., description=arg.description)
+            annotation = str
+        else:
+            default = Field(None, description=arg.description)
+            annotation = str | None
+
+        fields[arg.name] = annotation, default
+
+    return create_model(prompt.name, __base__=MCPClientTool, **fields)
+
+
+class MCPClient:
+    def __init__(self, session: ClientSession) -> None:
+        self.session: ClientSession = session
+
+    async def list_resources(self) -> list[Resource]:
```

Comment:
Here, the `type` of the mcp library is returned as it is. it would be ideal to define a mcp type specifically for mirascope, but I don't know how valuable this would be in the current situation, so this is the implementation.
What do you think?

Reply:
hmm yeah I think using the mcp lib types for now is good unless we have a clear reason to wrap